pisa/stages/osc/pi_globes.py

```diff
@@ -0,0 +1,298 @@
+                 debug_mode=None,
+                 input_specs=None,
+                 calc_specs=None,
+                 output_specs=None,
+                 detector_depth=2.*ureg.km,
+                 prop_height=20.*ureg.km
+                ):
+
+        expected_params = (
+                           'theta12',
+                           'theta13',
+                           'theta23',
+                           'deltam21',
+                           'deltam31',
+                           'deltam41',
+                           'theta24',
+                           'theta34',
+                           'deltacp',
+                          )
+
+        input_names = ()
+        output_names = ()
+
+        # what are the keys used from the inputs during apply
+        input_apply_keys = ('weights',
+                            'sys_flux',
+                           )
+        # what are keys added or altered in the calculation used during apply
+        output_calc_keys = ('prob_e',
+                            'prob_mu',
+                            'prob_nonsterile',
+                           )
+        # what keys are added or altered for the outputs during apply
+        output_apply_keys = ('weights',
+                      )
+
+        # init base class
+        super(pi_globes, self).__init__(data=data,
+                                        params=params,
+                                        expected_params=expected_params,
+                                        input_names=input_names,
+                                        output_names=output_names,
+                                        debug_mode=debug_mode,
+                                        input_specs=input_specs,
+                                        calc_specs=calc_specs,
+                                        output_specs=output_specs,
+                                        input_apply_keys=input_apply_keys,
+                                        output_calc_keys=output_calc_keys,
+                                        output_apply_keys=output_apply_keys,
+                                       )
+
+        assert self.input_mode is not None
+        assert self.calc_mode is not None
+        assert self.output_mode is not None
+
+        self.layers = None
+        self.osc_params = None
+        self.earth_model = earth_model
+        self.globes_wrapper = globes_wrapper
+        self.detector_depth = detector_depth
+        self.prop_height = prop_height
+
+        self.globes_calc = None
+        # This does nothing for speed, but just allows one to use numpy broadcasting on the
+        # arguments of the function. The internal implementation is basically a for-loop.
+        # The signature is chosen so that the function for a single event expects an array
+        # of n layer distances and densities.
+        self.calc_prob_e_mu = np.vectorize(self.calc_prob_e_mu, signature='(),(),(),(n),(n)->(),()')
```

Comment:
The function returned by numpy.vectorize has the same docstring as the original function unless specified otherwise, see https://docs.scipy.org/doc/numpy/reference/generated/numpy.vectorize.html
I can just add an underscore to the original function, so that `self.calc_prob_e_mu = np.vectorize(self._calc_prob_e_mu, ...) `.

Reply:
If that's the case, then my gripe is minor if not irrelevant. I wish there were a way to do the vectorization outside __init__, but I don't see anything straightforward, so feel free to keep the code as you have it.